### DISCUSSION TITLE
Some issues when using simmer that puzzle me regarding: get_mon_arrivals / clone / now()

### DISCUSSION THREAD

**Discussion by vfso** [r-simmer/simmer#291] [Q&A]
Good afternoon,

I am working on a simulation model using simmer. I am really happy with this great package!
So far, everything I want to model is possible. But every now and then I run into some issues that puzzle me.
These issues are the following:

1. During the testing of my model, I ran into the situation where the information obtained from get_mon_arrivals includes the same arrival twice (not clones). One record where the start_time and end_time are available, and another identical record with the only difference that the end_time is missing (NA). I have solved this by simply removing the duplicate record with the missing end_time before processing the results.
The strange thing is that this issue disappeared when I made some (seemingly unrelated) changes to the model. So unfortunately I cannot reproduce this issue and provide code that caused it. But before I remove the code lines that remove the duplicate records from get_mon_arrivals, I was wondering if anyone has encountered this before.

2. The second issue is actually a question. In my model there are processes that clone the arrival. These cloned arrivals then continue in the process as separate arrivals and will not be merged again with the original arrival. I have marked the clones by keeping track of an attribute "clonenr" (0 = original, 1 = clone 1, 2 = clone 2 etc.) hoping that I could use this attribute to assign the records obtained from get_mon_arrivals to the correct clone (combined with info from get_mon_attributes). I was not able to get this done and I think it is not possible to do this.
So instead, I created my own output data with information per arrival/clone which is filled during the simulation. This works fine, but of course adds to the time needed to run the model.
So I wanted to check whether my assumption is indeed correct, that records from get_mon_arrivals cannot be assigned to the right clone/ original arrival.

3. Somehow, the now() function sometimes returns an incorrect simulation time. This does not always happen, but I was able to create a small example (derived from the banking example) where this happens. 

```r
bank <- simmer("bank")

customerA <- function(s)
{
 path = trajectory("Customer's path") %>%
               log_(paste0("Tnow_A = ", now(s))) %>%
               log_("Here I am") %>%
               timeout(4) %>%
               log_("I must leave")
 return(path)
}

customerB <- function(s)
{
  join(trajectory() %>% 
           log_(paste0("Tnow_0 = ", now(s))),
         customerA(s))
}

add_generator(bank, "Customer", customerB(bank), at(5, 8))
bank %>% run(until = 50)
```

The output of the code is as follows:

```r
> bank %>% run(until = 50)
5: Customer0: Tnow_0 = 0
5: Customer0: Tnow_A = 0
5: Customer0: Here I am
8: Customer1: Tnow_0 = 0
8: Customer1: Tnow_A = 0
8: Customer1: Here I am
9: Customer0: I must leave
12: Customer1: I must leave
simmer environment: bank | now: 12 | next: 
{ Monitor: in memory }
{ Source: Customer | monitored: 1 | n_generated: 2 }
```

In the output, the now() function returns 0 for the simulation time, which is incorrect.
Does anyone know what is going on here?

Thanks for your help!

**Comment by Enchufa2** [comment]
1. You get the `NA`s when you specify `ongoing=TRUE` in the `get_mon_arrivals()` method. Basically, this option allows you to get the info of arrivals that didn't finish yet when you stop the simulation at any time.
2. Clones are not identified by design. If you want different identifiers, it's probably better to trigger a new arrival from a separate generator via the [`when_activated()`](https://r-simmer.org/reference/generators.html) helper.
3. See the section about [dynamic arguments](https://r-simmer.org/articles/simmer-03-trajectories.html#dynamic-arguments) in the documentation. Your code runs `now(s)` once when you define the trajectory, and the simulation time is zero at that point. What you want is to dynamically run this expression as the simulation advances, and this can be achieved with a function:

``` r
library(simmer)

bank <- simmer("bank")

customerA <- function(s) {
  path = trajectory("Customer's path") %>%
    log_(function() paste0("Tnow_A = ", now(s))) %>%
    log_("Here I am") %>%
    timeout(4) %>%
    log_("I must leave")
  return(path)
}

customerB <- function(s) {
  join(trajectory() %>% 
         log_(function() paste0("Tnow_0 = ", now(s))), customerA(s))
}

add_generator(bank, "Customer", customerB(bank), at(5, 8))
#> simmer environment: bank | now: 0 | next: 0
#> { Monitor: in memory }
#> { Source: Customer | monitored: 1 | n_generated: 0 }
bank %>% run(until = 50)
#> 5: Customer0: Tnow_0 = 5
#> 5: Customer0: Tnow_A = 5
#> 5: Customer0: Here I am
#> 8: Customer1: Tnow_0 = 8
#> 8: Customer1: Tnow_A = 8
#> 8: Customer1: Here I am
#> 9: Customer0: I must leave
#> 12: Customer1: I must leave
#> simmer environment: bank | now: 12 | next: 
#> { Monitor: in memory }
#> { Source: Customer | monitored: 1 | n_generated: 2 }
```

**Reply by vfso** [reply]
Thank you very much for your quick reply. Very much appreciated!
Regarding the issues:

1. Yes, I am aware of the option ongoing = TRUE and its functionality. But what I saw was that the same arrival appeared in the output twice. These two records did not refer to clones but to one and the same arrival entity. The records were identical with the only difference the end_time being NA for one of them. I do realize that it is difficult to look into this since I don't have code that replicates this issue.
So as soon as I encounter this again I will get back at this.
2. Thanks for the tip. I will look into the when_activated() function.
3. Ah, indeed... I missed this one since the code did not throw an error like when using e.g. get_attribute() without a function.

I can see from the discussions that you are really involved in the development and use of the simmer package. Thank you so much for this! I am trying to promote the use of DES in our organization which is quite a challenge (especially when using open source tools instead of a tool like Simio). But it certainly helps that simmer is actively used, discussed and supported on this forum. Thanks!

**Reply by Enchufa2** [reply]
Glad to hear it! I keep tracking bugs down, and simmer receives one release per year or so. Many of the features were developed as a response to discussions and requests in the mailing list. 

Of course, please let me know if you reproduce the issue. Thanks!
